Fixes #8433 - non-apostrophed ('...') url with multiple params support in cURL import

SahilChoudhary22 opened the pull request:
- Fixes https://github.com/postmanlabs/postman-app-support/issues/8843
- Added few more test cases to support regex changes.

umeshp7 commented on test/conversion.test.js:
```diff
@@ -594,4 +594,115 @@ describe('Curl converter should', function() {
+      }, function (err, result) {
+        expect(result.result).to.equal(true);
+        expect(result.output.length).to.equal(1);
+        expect(result.output[0].type).to.equal('request');
```
The test has to expect the right information you are expecting inside the request as well. For eg. the url in this case.

umeshp7 commented on assets/shell-quote.js:
```diff
@@ -25,6 +25,7 @@ var META = '|&;()<> \\t';
 var BAREWORD = '(\\\\[\'"' + META + ']|[^\\s\'"' + META + '])+';
```
@SahilChoudhary22 This is an external asset that we have added to the module. We will have to be completely sure before we make any changes to it.

umeshp7 replied:
@SahilChoudhary22 This is an external asset that we have added to the module. We will have to be completely sure before we make any changes to it.

umeshp7 commented on test/conversion.test.js:
```diff
@@ -594,4 +594,115 @@ describe('Curl converter should', function() {
+        -H "DeviceId: b147ee89" \\
+        -H "pid: bdb5748e" \\
+        -H "sid: 5096fd6e" \\
+        -H "x-features: FEATURE1_ENABLED" \\
```
We can remove some of the unwanted headers etc. here and use the smallest version which reproduces the issue.

umeshp7 replied:
We can remove some of the unwanted headers etc. here and use the smallest version which reproduces the issue.

umeshp7 commented on assets/shell-quote.js:
```diff
@@ -25,6 +25,7 @@ var META = '|&;()<> \\t';
 var BAREWORD = '(\\\\[\'"' + META + ']|[^\\s\'"' + META + '])+';
 var SINGLE_QUOTE = '"((\\\\"|[^"])*?)"';
 var DOUBLE_QUOTE = '\'((\\\\\'|[^\'])*?)\'';
+var HTTP_URL_MATCHER  = "http[^ ]+";
```
@SahilChoudhary22 can we check the shell-quote npm module, if there are any open issues around this or if there is any handling already for this?

umeshp7 commented on assets/shell-quote.js:
```diff
@@ -25,6 +25,7 @@ var META = '|&;()<> \\t';
 var BAREWORD = '(\\\\[\'"' + META + ']|[^\\s\'"' + META + '])+';
 var SINGLE_QUOTE = '"((\\\\"|[^"])*?)"';
 var DOUBLE_QUOTE = '\'((\\\\\'|[^\'])*?)\'';
+var HTTP_URL_MATCHER  = "http[^ ]+";
```
@SahilChoudhary22 can we check the shell-quote npm module, if there are any open issues around this or if there is any handling already for this?

umeshp7 replied:
The url could simply be www.google.com?foo=bar&foo=baz or google.com?foo=bar&foo=baz and these cases aren't covered with this.

SahilChoudhary22 commented:
- Rolled back the changes in `assets/shell-quote.js` to avoid modification in an external library. 
- Made changes in `src/lib.js` and added cURL string preprocessing to handle incorrectly formed input.
- Compacted the test cases, removed unnecessary headers. Added some more assertions in each test for proper testing.

umeshp7 commented on test/conversion.test.js:
```diff
@@ -594,4 +594,177 @@ describe('Curl converter should', function() {
+        -H "User-Agent: Dalvik/2.1.0 (Linux; U; Android 10; SM-A705FN Build/QP1A.190711.020) Test/4.6.4.459" \\
+        -H "Authorization: bearer XXX" \\
+        -H "Content-Type: application/json" \\
+        "test.com/?bar=1&baz=2"`
```
let's add a test with more than 2 query params.

umeshp7 commented on src/lib.js:
```diff
@@ -466,7 +466,11 @@ var program,
+        // [Github #8843] - RegEx to fix malformed cURLs with unquoted multi-param URLs
+        const multiParamUrlRegEx = /[^ ]+\.[^ ]+&[^ ]+/gm;
+
+        var cleanedCurlString = curlString.replace(multiParamUrlRegEx, `'${curlString.match(multiParamUrlRegEx)}'`),
```
Will single quotes always work?

umeshp7 commented on src/lib.js:
```diff
@@ -466,7 +466,11 @@ var program,
 
-        var sanitizedArgs = this.sanitizeArgs(curlString),
+        // [Github #8843] - RegEx to fix malformed cURLs with unquoted multi-param URLs
+        const multiParamUrlRegEx = /[^ ]+\.[^ ]+&[^ ]+/gm;
```
What will happen if we have a URL like this > 
`https://postman-echo.com/get?foo1=bar1&` 

@SahilChoudhary22

umeshp7 commented:
We should add additional handling for trailing `&` in the URL which is not accounted for in the current regex.

umeshp7 commented:
We should safe check at `arg.startsWith('$')`. And throw appropriate errors where necessary. 

If this `&` is still encountered after the handling like:`curl https://postman-echo.com/post -d foo1=bar1&foo2=bar2` these are clearly invalid. Then we should respond with an error saying "Error parsing curl: <message>"

umeshp7 commented on test/conversion.test.js:
```diff
@@ -594,4 +594,269 @@ describe('Curl converter should', function() {
+      convert({
+        type: 'string',
+        data: `curl -X GET \\
+        -H "User-Agent: Dalvik/2.1.0 (SM-A705FN Build/QP1A.190711.020)" \\
+        -H "Authorization: bearer XXX" \\
+        -H "Build: 4.6.4.459" \\
+        -H "Platform: Android" \\
+        -H "Accept-Language: tr-TR" \\
+        -H "Content-Type: application/json" \\
```
Do we need all these headers?

umeshp7 commented:
@SahilChoudhary22 Looks good to me. 
Let's merge this and we can test this one beta.
